### DISCUSSION TITLE
Red Dead Redemption

### DISCUSSION THREAD

**Discussion by bomabomaboma** [xenia-canary/game-patches#8] [General]
@illusion0001 
If you set the stream range of all non-tree plants to a static value, the grass-streaming issue is solved at a modest cost to overall draw distance.

I'm still building new shaders because I changed values in two of compiled shader files, but my framerate is 5-25% better depending on the area. The lower overall draw distance just looks like an environmental setting was lowered in a PC game, without sudden flashing in and out of vast swaths of grassland.

Is there interest in pursuing this further?

DoF values dynamically change between shaders, weather, time of day, and area of the map. I'm not so sure that can be easily toggled.

I spent 90 minutes picking up bread crumbs, looking for motionblur, but I fear it's hidden in an encrypted script file for which there is no available tool to view.

**Comment by bomabomaboma** [comment]
I'll get some capture software to demonstrate the difference.

**Comment by bomabomaboma** [comment]
I'm trying to debug Xenia for RDR, and am getting overwhelmed with 10GB of
```
!> F8000928 XMA: Write to unknown register (0601): 00000002
!> F8000928 XMA: Write to unknown register (0601): 00000003
!> F80009E4 XMA: Read from unhandled register (0600, CONTEXT_ARRAY_ADDRESS)
```
within 20 minutes of gameplay.
Is there a known method to building Xenia so that these unknown register alerts are removed from the log?

**Reply by illusionyy** [reply]
Try this https://github.com/xenia-canary/xenia-canary/releases/tag/6a2570595

**Comment by bomabomaboma** [comment]
@illusion0001 @emoose 

Found a potential lead on fixing RDR not allowing campfires, but I don't understand these PPC opcodes. This syntax bears zero resemblance to other Rockstar games.

There is a trigger to force your horse to spawn at your newly built campfire, and we can easily create a patch for infinite campfires.

```
spush	spush	[0x33] 'Got a request to create a camp, lets see if we can'
native	native	0245 => 69:$FD25473E, (1):0
getf	0
pframe1	3
call2h1	call 	#10BC0
jfr2	69 => #45
getf	3
jfr2	61 => #3D
iimm0	ipush	0
getf	0
psetimm1	40
fimm0	fpush	0.00
fimm0	fpush	0.00
fimm0	fpush	0.00
iimm3	ipush	3
getf	0
fromstack	
spush	spush	[0x1C] 'We can, lets set some bits!'
native	native	0245 => 69:$FD25473E, (1):0
getf	0
iaddimm1	32
iimm2	ipush	2
call2	call 	#21F5
getf	0
iaddimm1	32
iimm1	ipush	1
call2h1	call 	#17C62
jr2	697 => #2B9
iimm0	ipush	0
getf	0
psetimm1	40
getf	0
pgetimm1	32
ipush1	ipush	16
call2	call 	#1150
jfr2	82 => #52
spush	spush	[0x1D] 'We can't, one already exists'
native	native	0245 => 69:$FD25473E, (1):0
spush	spush	[0x10] 'Camp_Not_EXISTS'
ipush	ipush	1092616192
iimm1	ipush	1
iimm0	ipush	0
iimm2	ipush	2
iimm1	ipush	1
iimm0	ipush	0
call2	call 	#48C5
spush	spush	[0x22] 'We can't, no good location nearby'
native	native	0245 => 69:$FD25473E, (1):0
spush	spush	[0x13] 'Camp_Not_PROBEFAIL'
ipush	ipush	1092616192
iimm1	ipush	1
iimm0	ipush	0
iimm2	ipush	2
iimm1	ipush	1
iimm0	ipush	0
call2	call 	#48C5
spush	spush	[0x13] 'We can't, in town!'
native	native	0245 => 69:$FD25473E, (1):0
spush	spush	[0xE] 'Camp_Not_TOWN'
ipush	ipush	1092616192
iimm1	ipush	1
iimm0	ipush	0
iimm2	ipush	2
iimm1	ipush	1
iimm0	ipush	0
call2	call 	#48C5
spush	spush	[0x19] 'We can't, unknown reason'
native	native	0245 => 69:$FD25473E, (1):0
```

**Reply by emoose** [reply]
FYI the FD25473E value here seems to be LOG_WARNING, based on the natives listed at https://pastebin.com/HMcNtPHK, I wonder if that could be hooked to call DbgPrint with each warning msg...

I'd guess the `call2h1	call 	#10BC0` is calling some other script func at 0x10BC0 that returns whether campfire is allowed, then seems to use jfr2 to jump to other sections depending on what 0x10BC0 returns. 

If you can disassemble 0x10BC0 that might tell us what natives it uses to check that. I don't really know what opcodes these are though, definitely not PPC, some kind of scripting bytecode?
E: https://forum.xentax.com/viewtopic.php?p=71624#p71624 has a list of these opcodes and a small description for some of them, not sure if that's the best list available though.

IIRC there was some talk about disassembling/fixing RDR etc before on the canary discord, not sure how much progress was made there though.

**Reply by illusionyy** [reply]
Don't think this is PPC, could be Rockstar compiled scripting language.
Do you have a save near campfire?
Edit: try to disasm 0x10BC0 as emoose said. Might help.

**Comment by bomabomaboma** [comment]
EDIT: This is for Disc 1 of GOTY only.

No update on the campfire yet. Sorry, I've had a migraine and haven't been able to look at a bright screen much.

I'm including an .xdelta for patching rdr2_layer0.rpf that removes all grass rendering, for your testing purposes. (I have the GOTY edition, this will not work on the original 2010 release, there are too many differences in the archives.)

Setting the clip area of all grasses to 0.0 0.0 0.0 0.0 in /root/grassres/patches.txt
![default](https://user-images.githubusercontent.com/81534189/115096837-ca0fed80-9edb-11eb-8270-e26ea2b54091.jpg)

All the grass .dds will still occupy the texture cache, but grass is no longer rendered. Framerate is about the same but frametime fluctuations occur less often, and overall the image is less distracting.

Not sure if it can be made into a Xenia patch, patches.txt is mentioned in the .xex, but deleting it altogether crashes the game.
[rdr-grass-removed.zip](https://github.com/xenia-canary/game-patches/files/6328412/rdr-grass-removed.zip)

![example](https://user-images.githubusercontent.com/81534189/115097172-71d9eb00-9edd-11eb-864a-883767a96b9e.jpg)

**Comment by bomabomaboma** [comment]
Unrelated Update:
I've seen several people request a patch that disables motion blur. I know that there are several "motionblur" strings in the .xex, but I believe that those are referencing a minimal per-object motion blur that actually helps encourage a smooth experience below 60fps. 

I cannot find any evidence of camera motion blur. In fact, what's actually happening is a severe, overbearing DoF that is enabled when you pan the camera above a certain threshold.

DoF Enabled: (Notice the total lack of blur on objects in the foreground.
![panning-dof](https://user-images.githubusercontent.com/81534189/115104203-b9776b80-9f0b-11eb-9258-78b2485504fc.png)

Dof Disabled:
![panning-dof2](https://user-images.githubusercontent.com/81534189/115104219-cf852c00-9f0b-11eb-9abc-c3f2fd8a4ec1.png)


I'll elaborate tomorrow and send an xdelta along with helpful save-states.
It's late here.

**Reply by illusionyy** [reply]
Skipped check for grass occlusion entirely. https://github.com/xenia-canary/game-patches/commit/ab34a13fee8dd43f700791807f7375a0fd1f4c43
Edit: Motion Blur too https://github.com/xenia-canary/game-patches/commit/c818f709341d23f9ec6c5e97486886bc414aa8fb

**Reply by bomabomaboma** [reply]
Nice!

**Reply by illusionyy** [reply]
Dunno if it will work for your game though, since mine is original disc version (without undead nightmare on disc) so will probably needs to be ported

**Reply by bomabomaboma** [reply]
@illusion0001 
Tried the RDR patches with GOTY. Updated Media ID and Hash, they _were_ applied.
Neither grass nor blur patches work for either disc.

Perhaps show me a video of your process? As you did with GoW2.

**Reply by bomabomaboma** [reply]
![image](https://user-images.githubusercontent.com/81534189/115812649-8c610800-a3a6-11eb-8147-600b2ed454e1.png)
I haven't enabled my no-grass mod, obviously.

**Reply by illusionyy** [reply]
> Perhaps show me a video of your process? As you did with GoW2.

If you'd like to follow into a rabbit hole of ppc asm, sure 😹

**Reply by illusionyy** [reply]
Soo, it applied but the patches aren't working? Seems to be patching something else then. Good that it isn't crashing outright though

**Comment by FlintEastwood** [comment]
Hello guys,
is there a way to trick the game into a native 1080p resolution?

I'm asking, because with xenia's resolution scaling the rain will be invisible. That feels pretty weird.
![720p_Rain](https://github.com/xenia-canary/game-patches/assets/19540104/5c4b9b41-f464-417c-b07f-89a73284f15e)
![1440pResScale2x_Rain](https://github.com/xenia-canary/game-patches/assets/19540104/76cc93c9-df4d-4190-81d3-d47e0e5427a4)

**Reply by MaxM2099** [reply]
Use the 540p plus 4x msaa patch and 2x the resolution.

**Reply by MaxM2099** [reply]
My bad you said native. ATM the only way is to wait until you can change res in the toml file because their is an option just that it works in  very few games maybe none not sure.
